great_expectations/core/added_diagnostics.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import ClassVar, Tuple, Type
+
+from great_expectations.compatibility.typing_extensions import override
+from great_expectations.exceptions.exceptions import (
+    BatchDefinitionNotAddedError,
+    CheckpointNotAddedError,
+    CheckpointRelatedResourcesNotAddedError,
+    ExpectationSuiteNotAddedError,
+    ResourceNotAddedError,
+    ResourcesNotAddedError,
+    ValidationDefinitionNotAddedError,
+    ValidationDefinitionRelatedResourcesNotAddedError,
+)
+
+
+@dataclass
+class AddedDiagnostics:
+    """
+    Wrapper around a list of errors; used to determine if a resource has been added successfully.
+
+    Note that some resources may have dependencies on other resources - in order to be considered
+    "added", the root resource and all of its dependencies must be added successfully.
+    For example, a Checkpoint may have dependencies on ValidationDefinitions, which may have
+    dependencies on ExpectationSuites and BatchDefinitions.
+
+    GX requires that all resources are added successfully before they can be used to prevent
+    unexpected behavior.
+    """
+
+    errors: list[ResourceNotAddedError]
+
+    @property
+    def is_added(self) -> bool:
+        return len(self.errors) == 0
+
+    @abstractmethod
+    def raise_for_error(self) -> None:
+        """
```

Comment:
So this will either raise or potentially return None?

Reply:
Yup akin to `requests.Response.raise_for_status`

Reply:
could do something like:
```
if not diagnostics.is_added:
    raise diagnostics.exception()
```
but the conditional check will be the same with every use